crates/services/txpool_v2/src/collision_manager/mod.rs

```diff
@@ -16,46 +16,49 @@ use fuel_core_types::{
     services::txpool::PoolTransaction,
 };
 
-use crate::error::{
-    CollisionReason,
-    Error,
+use crate::{
+    error::{
+        CollisionReason,
+        Error,
+    },
+    storage::StorageData,
 };
 
 pub mod basic;
+pub mod collision;
+
+pub trait Collision<StorageIndex> {
+    /// Returns the instigator of the collision.
+    fn tx(&self) -> &PoolTransaction;
+
+    /// Returns the transactions that collide with the given transaction.
+    fn colliding_transactions(&self) -> &HashMap<StorageIndex, Vec<CollisionReason>>;
+
+    /// Converts the collision into a collection of storage indexes.
+    fn into_colliding_storage(self) -> Vec<StorageIndex>;
+}
 
 pub trait CollisionManager {
     /// Storage type of the collision manager.
     type Storage;
     /// Index that identifies a transaction in the storage.
     type StorageIndex;
+    /// The collision type returned by the collision manager.
+    type Collision<'a>: Collision<Self::StorageIndex>;
 
-    /// Collect the transaction that collide with the given transaction.
-    /// Returns a list of storage indexes of the colliding transactions and the reason of the collision.
-    fn collect_colliding_transactions(
-        &self,
-        transaction: &PoolTransaction,
-    ) -> Result<HashMap<Self::StorageIndex, Vec<CollisionReason>>, Error>;
-
-    /// Determine if the collisions allow the transaction to be stored.
-    /// Returns the reason of the collision if the transaction cannot be stored.
-    fn can_store_transaction(
+    /// Finds collision for the transaction. Returns `Self::Collision`, if find any.
+    fn find_collision<'a>(
```

Comment:
it's strange to me that a "collision" can in fact collide with multiple transactions and collision with one tx has nothing to do with collision with another one. Is there a reason why you don't like "collisions" ?

Reply:
Renamed to has "s"=)